src/rtc.c

```diff
@@ -0,0 +1,341 @@
+ *
+ * @param[in]   bcd
+ *              binary-coded decimal number
+ *
+ * @return the decoded number
+ */
+static uint8_t bcd_to_byte(uint8_t bcd)
+{
+    uint8_t hi = (bcd & 0xF0) >> 4;
+    uint8_t lo = bcd & 0x0F;
+    return (hi * 10) + lo;
+}
+
+/**
+ * @brief Encode a binary-coded decimal number.
+ *
+ * @param[in]   byte
+ *              number to encode
+ *
+ * @return the encoded binary-coded decimal number
+ */
+static uint8_t byte_to_bcd(uint8_t byte)
+{
+    byte %= 100;
+    return ((byte / 10) << 4) | (byte % 10);
+}
+
+/**
+ * @brief Read the identifier and status of the real-time clock.
+ * 
+ * @param[out]  identifier
+ *              Destination pointer for the RTC identifier byte (or NULL).
+ * @param[out]  status 
+ *              Destination pointer for the RTC status byte (or NULL).
+ */
+void joybus_rtc_status( uint8_t * identifier, uint8_t * status )
+{
+    static const uint64_t input[8] =
+    {
+        0x00000000ff010306,
+        0xfffffffe00000000,
+        0,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    controller_exec_pif( input, output );
+
+    uint8_t * recv_buf = (uint8_t *)&output[1];
+
+    if( identifier != NULL ) *identifier = recv_buf[1];
+    if( status != NULL ) *status = recv_buf[2];
+}
+
+/**
+ * @brief Probe the RTC interface on the cartridge.
+ *
+ * Some flash carts require the RTC to be explicitly enabled and
+ * will not report the RTC as present by default.
+ * 
+ * Some emulators do not support the RTC correctly or at all.
+ *
+ * @return whether the real-time clock was detected on the cartridge.
+ */
+bool joybus_rtc_present( void )
+{
+    uint8_t identifier;
+    joybus_rtc_status( &identifier, NULL );
+    return identifier == JOYBUS_RTC_IDENTIFIER;
+}
+
+/**
+ * @brief Read the status of the real-time clock.
+ * 
+ * The RTC will be busy when the control block is in
+ * #JOYBUS_RTC_CONTROL_MODE_SET and is ready to be written to.
+ * 
+ * @return whether the Joybus RTC is busy
+ */
+bool joybus_rtc_busy( void )
+{
+    uint8_t status;
+    joybus_rtc_status( NULL, &status );
+    return status == JOYBUS_RTC_STATUS_BUSY;
+}
+
+/**
+ * @brief Read the control block data from the real-time clock.
+ * 
+ * The control block contains the bits for:
+ * #JOYBUS_RTC_CONTROL_MODE_SET when writing to the RTC.
+ * #JOYBUS_RTC_CONTROL_BLOCK1_WRITE_PROTECTED is not used.
+ * #JOYBUS_RTC_CONTROL_BLOCK2_WRITE_PROTECTED when not writing to the RTC.
+ * 
+ * @return the RTC control block data
+ */
+uint16_t joybus_rtc_read_control( void )
+{
+    static const uint64_t input[8] =
+    {
+        0x0000000002090700,
+        0xffffffffffffffff,
+        0xfffe000000000000,
+        0,
+        0,
+        0,
+        0,
+        1
+    };
+    static uint64_t output[8];
+
+    controller_exec_pif( input, output );
+
+    uint8_t * recv_buf = (uint8_t *)&output[1];
+
+    return (((uint16_t)recv_buf[1]) << 8) | recv_buf[0];
```

Comment:
This doesn't seem right to me, but appears to be correct based on the UltraPIF implementation: https://github.com/jago85/UltraPIF_MCU/blob/3b185ce4b2afd8d0489c992061ab43a8f39b064c/Src/main.c#L584-L585

@jago85 Are you sure that the control block bytes are swapped when reading vs writing?

Reply:
You're right. This seems to be a bug in my implementation. I'll fix this and check if Animal Forest is happy.

Reply:
I've checked with a real RTC. Of course, the byte order for reading and writing is the same. I'll fix the Read Control Register feature in my implementation. So I will keep interpreting the data as big endian uint16_t. Could also be the other way around, but this is the simplest change for me.
Animal Forest is also accepting the changes. It seems like the Control Register is only read to perform a Read-Modify-Write operation.